.github/workflows/main.yml

```diff
@@ -1,193 +1,135 @@
-      with:
-        submodules: 'true'
-    - uses: conda-incubator/setup-miniconda@d2e6a045a86077fb6cad6f5adf368e9076ddaa8d  # v3.1.0
-      with:
-        miniforge-variant: Miniforge3
-        miniforge-version: latest
-        activate-environment: cpp_test
-        environment-file: tests/ci_build/conda_env/cpp_test.yml
-        use-mamba: true
-    - name: Display Conda env
-      run: |
-        conda info
-        conda list
-
-    - name: Build and install XGBoost static library
-      run: |
-        mkdir build
-        cd build
-        cmake .. -DBUILD_STATIC_LIB=ON -DCMAKE_INSTALL_PREFIX=$CONDA_PREFIX -GNinja
-        ninja -v install
-        cd -
-    - name: Build and run C API demo with static
-      run: |
-        pushd .
-        cd demo/c-api/
-        mkdir build
-        cd build
-        cmake .. -GNinja -DCMAKE_PREFIX_PATH=$CONDA_PREFIX
-        ninja -v
-        ctest
-        cd ..
-        rm -rf ./build
-        popd
-
-    - name: Build and install XGBoost shared library
-      run: |
-        cd build
-        cmake .. -DBUILD_STATIC_LIB=OFF -DCMAKE_INSTALL_PREFIX=$CONDA_PREFIX -GNinja -DPLUGIN_FEDERATED=ON -DGOOGLE_TEST=ON
-        ninja -v install
-        ./testxgboost
-        cd -
-    - name: Build and run C API demo with shared
-      run: |
-        pushd .
-        cd demo/c-api/
-        mkdir build
-        cd build
-        cmake .. -GNinja -DCMAKE_PREFIX_PATH=$CONDA_PREFIX
-        ninja -v
-        ctest
-        popd
-        ./tests/ci_build/verify_link.sh ./demo/c-api/build/basic/api-demo
-        ./tests/ci_build/verify_link.sh ./demo/c-api/build/external-memory/external-memory-demo
-
-  cpp-lint:
-    runs-on: ubuntu-latest
-    name: Code linting for C++
+        include:
+          - description: single-gpu
+            container: xgb-ci.gpu
+            suite: gpu
+            runner: linux-amd64-gpu
+            artifact_from: build-cuda
+          - description: multiple-gpu
+            container: xgb-ci.gpu
+            suite: mgpu
+            runner: linux-amd64-mgpu
+            artifact_from: build-cuda
     steps:
-    - uses: actions/checkout@a5ac7e51b41094c92402da3b24376905380afc29 # v4.1.6
-      with:
-        submodules: 'true'
-    - uses: actions/setup-python@0b93645e9fea7318ecaed2b359559ac225c90a2b # v5.3.0
-      with:
-        python-version: "3.10"
-        architecture: 'x64'
-    - name: Install Python packages
-      run: |
-        python -m pip install wheel setuptools cmakelint cpplint==1.6.1 pylint
-    - name: Run lint
-      run: |
-        python3 tests/ci_build/lint_cpp.py
-        sh ./tests/ci_build/lint_cmake.sh
+      # Restart Docker daemon so that it recognizes the ephemeral disks
+      - run: sudo systemctl restart docker
+      - uses: actions/checkout@v4
+        with:
+          submodules: "true"
+      - name: Fetch container from cache
+        run: bash ops/docker_build.sh ${{ matrix.container }}
```

Comment:
> am I right that there's only a pull happening here, not a new build and push?

Yes (due to the use of Elastic Container Registry as a caching layer)

> why is this necessary at all?

Currently, `docker_run.py` does not automatically pull the container from ECR. I can update the script to make it do so.

Reply:
> *I can update the script to make it do so*

You don't need to have a separate `docker pull` step. When `docker_run.py` runs `docker run ${image}`, if  `${image}` does not exist, it will be pulled automatically.

Reply:
> if ${image} does not exist, it will be pulled automatically.

Right now it doesn't, because `docker run` would attempt to pull `xgb-ci.gpu` (for example) with no prefix. Adding a prefix (in this case `492475357299.dkr.ecr.us-west-2.amazonaws.com/`) will do the trick.

Reply:
Ohhhhh ok I see. That could be avoided by passing a fully-qualified image URI to `docker run`, of the form like:

```text
${REGISTRY}/${REPO}:${TAG}
```

e.g.

```text
492475357299.dkr.ecr.us-west-2.amazonaws.com/xgb-ci.gpu:main
```

There are some notes about that in the EKS docs, just as one example: https://docs.aws.amazon.com/emr/latest/EMR-on-EKS-DevelopmentGuide/docker-custom-images-tag.html